### PR TITLE
feat(tracing): use new common tracing config from from agent response

### PR DESCRIPTION
**Pull request by basti1302**
~Note: one assertion in the tests is failing, I'll look into this tomorrow/Friday/early next week.~

With the upcoming Node.js discovery plug-in version 1.2.17, common tracing
configuration (including extra http headers and the new Kafka tracing config)
will be available under the "tracing" attribute in the agent's response.

Use the values from agentResponse.tracing over the legacy attributes for
extraHeaders. Also, apply the new Kafka tracing config (correlation
disabled/enabled, header format) to our Kafka instrumentations.

refs 93724

### CODE REVIEW

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js [hunk 1]
```diff
@@ -240,13 +241,23 @@ mochaSuiteFn('tracing/messaging/node-rdkafka', function () {
...
+          expect(span.t).to.be.a('string');
+        }
+      },
+      span => {
+        if (parent) {
+          expect(span.p).to.equal(parent.s);
+        } else {
+          expect(span.p).to.not.exist;
+        }
+      },
```
Comment (basti1302) [inline packages/collector/test/tracing/messaging/node-rdkafka/test.js:260]: This did not actually check for span.p to be undefined in the previous version.

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js [hunk 2]
```diff
@@ -291,6 +302,154 @@ mochaSuiteFn('tracing/messaging/node-rdkafka', function () {
     ]);
   }
 
+  describe('tracing enabled, header format string', function () {
```
Comment (basti1302) [inline packages/collector/test/tracing/messaging/node-rdkafka/test.js:305]: We did not have tests with different header formats for rdkafka before. `binary` is not used, so it is fine that that is missing, but a test for `string` was missing so I added it.

File: packages/collector/test/tracing/messaging/node-rdkafka/test.js [hunk 3]
```diff
@@ -291,6 +302,154 @@ mochaSuiteFn('tracing/messaging/node-rdkafka', function () {
...
+        const httpEntry = verifyHttpRootEntry({ spans, apiPath, pid: String(producerControls.getPid()) });
+        const kafkaExit = verifyRdKafkaExit(producerControls, spans, httpEntry, null, false);
+        verifyHttpExit({ spans, parent: httpEntry, pid: String(producerControls.getPid()) });
+        const kafkaEntry = verifyRdKafkaEntry(consumerControls, spans, kafkaExit, null, false);
+        verifyHttpExit({ spans, parent: kafkaEntry, pid: String(consumerControls.getPid()) });
+      });
+    });
+  });
+
+  describe('tracing enabled, but trace correlation disabled', function () {
```
Comment (basti1302) [inline packages/collector/test/tracing/messaging/node-rdkafka/test.js:379]: This case had also been missing previously.

File: packages/core/src/tracing/constants.js [hunk 4]
```diff
@@ -28,6 +28,11 @@ exports.kafkaTraceIdHeaderNameString = 'X_INSTANA_T';
 exports.kafkaSpanIdHeaderNameString = 'X_INSTANA_S';
 exports.kafkaTraceLevelHeaderNameString = 'X_INSTANA_L_S';
 
+exports.kafkaTraceCorrelationDefault = true;
+// Before we start phase 1 of the migration, 'binary' will be the default value. With phase 1, we will move to 'both',
+// with phase 2 it will no longer be configurable and will always use 'string'.
+exports.kafkaHeaderFormatDefault = 'binary';
```
Comment (basti1302) [inline packages/core/src/tracing/constants.js:34]: Extracted the default from the two instrumentation modules that are using it.

File: packages/core/src/tracing/index.js [hunk 5]
```diff
@@ -23,8 +23,6 @@ let automaticTracingEnabled = false;
 let cls = null;
 /** @type {import('../util/normalizeConfig').InstanaConfig} */
 let config = null;
-/** @type {Array.<string>} */
-let extraHeaders = [];
```
Comment (basti1302) [inline packages/core/src/tracing/index.js:27]: We kept a copy of the extraHeaders setting in this module but never used it for anything, so I removed it.

File: packages/core/src/tracing/index.js [hunk 6]
```diff
@@ -235,10 +240,9 @@ exports.getCls = function getCls() {
 };
 
 /**
- * @param {Array.<string>} _extraHeaders
+ * @param {Array.<string>} extraHeaders
  */
-exports.setExtraHttpHeadersToCapture = function setExtraHttpHeadersToCapture(_extraHeaders) {
-  extraHeaders = _extraHeaders;
```
Comment (basti1302) [inline packages/core/src/tracing/index.js:241]: This is where we wrote the obsolete extraHeaders (and never read it again).

File: packages/core/src/tracing/instrumentation/messaging/kafkaJs.js [hunk 7]
```diff
@@ -15,10 +15,8 @@ logger = require('../../../logger').getLogger('tracing/kafkajs', newLogger => {
   logger = newLogger;
 });
 
-let traceCorrelationEnabled = true;
-// Before we start phase 1 of the migration, 'binary' will be the default value. With phase 1, we will move to 'both',
-// with phase 2 it will no longer be configurable and will always use 'string'.
```
Comment (basti1302) [inline packages/core/src/tracing/instrumentation/messaging/kafkaJs.js:20]: comment was moved to the constants file.

File: packages/core/src/tracing/instrumentation/messaging/rdkafka.js [hunk 8]
```diff
@@ -16,36 +16,50 @@ logger = require('../../../logger').getLogger('tracing/rdkafka', newLogger => {
...
- * Kafka migration has not officially started yet
- * The option `headerFormat` is not officially documented yet.
- *
- * To ensure customers can use rdkafka instrumentation we set the
- * default to "both" here. That means we send both header formats
- * by default.
- *
- * Node.js Sender -> Node.js Receiver works
- * Node.js Sender -> Java Receiver does not works
- */
```
Comment (basti1302) [inline packages/core/src/tracing/instrumentation/messaging/rdkafka.js:31]: comment moved down to overrideKafkaHeaderFormat function

File: packages/core/test/tracing/index_test.js [hunk 9]
```diff
@@ -7,49 +7,77 @@
...
+    setKafkaTracingConfigStubRdKafka = sinon.stub(rdKafka, 'setKafkaTracingConfig');
   });
 
-  it('deactivate instrumentation via config', () => {
-    const config = { tracing: { disabledTracers: ['grpc', 'kafka'] } };
-    normalizeConfig(config);
+  beforeEach(() => {
+    // requiring tracing/index via proxyquire gives us a module in pristine state everytime, in particular with the
+    // tracingActivated flag reset.
+    tracing = proxyquire('../../src/tracing', {});
```
Comment (basti1302) [inline packages/core/test/tracing/index_test.js:40]: There were some issues with this test suite due to the fact that we re-used the same (stateful) instance of `require('../../src/tracing')` for all tests. In particular, the `activate` method was not executed, except for the first test in the suite.

File: packages/collector/test/announceCycle/unannounced_test.js [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+const { secrets, tracing } = require('@instana/core');
+const { constants } = tracing;
+
+let unannouncedState;
+let agentResponse;
+
+const pidStoreStub = {};
+const agentOptsStub = {};
+const tracingStub = sinon.stub(tracing);
+const secretsStub = sinon.stub(secrets);
```
Comment (kirrg001) [inline packages/collector/test/announceCycle/unannounced_test.js:24]: This won't work, because if you run the tests this code will be executed for **all tests**.
That means you would stub the whole core package for all tests.
The stub call needs to go into `before` or `beforeEach` - depends what you want to achieve :)

File: packages/collector/src/announceCycle/unannounced.js [hunk 11]
```diff
@@ -64,43 +91,106 @@ function tryToAnnounce(ctx, retryDelay = initialRetryDelay) {
...
+      traceCorrelation:
+        kafkaTracingConfigFromAgent['trace-correlation'] != null
+          ? kafkaTracingConfigFromAgent['trace-correlation']
+          : tracingConstants.kafkaTraceCorrelationDefault,
+      headerFormat:
+        kafkaTracingConfigFromAgent['header-format'] != null
+          ? kafkaTracingConfigFromAgent['header-format']
+          : tracingConstants.kafkaHeaderFormatDefault
+    };
+    tracing.setKafkaTracingConfig(kafkaTracingConfig);
```
Comment (kirrg001) [inline packages/collector/src/announceCycle/unannounced.js:182]: _Suggestion_:

The code to override the config from the agent for the target instrumentation is a bit hard to follow, but of course it existed before this PR :)

As far as I can see, we do not trace before the announceCycle is completed and reaches `agentReady` step, which calls `tracing.activate`.

In the `unannounced` step we try to call the core package (which I think is unneeded and too coupled) to tell the instrumentation that there is another config available, which is more important.

IMO the unannounced cycle step should write the agent config into the ctx object. And when the last step is reached we pass `tracing.activate(ctx.agentConfig)`. We then automatically pass the agent config into each instrumentation. Each instrumentation can pick the config it needs.

Advantage: 
- we do not have to iterate 3 times through all instrumentations (costs time)
- easier to read
- we don't need setExtraHttpHeadersToCapture, setKafkaTracingConfig or similar. If an instrumentation needs config from agent, it can just access the agentConfig in the activate fn

Happy to discuss and refactor afterwards.

File: packages/core/test/tracing/index_test.js (hunk 9 above)
Comment (kirrg001) [reply packages/core/test/tracing/index_test.js:43]: Maybe this is cased by [this line](https://github.com/instana/nodejs/pull/527/files#diff-e2d7b3fbf936c12cd8fa74f525607ea570aa81c0a5bc68de09c8ea6eeb0e95f0R23)?

**Review by kirrg001 — COMMENTED** [review]
LGTM 👍 Left some comments :)

File: packages/collector/src/announceCycle/unannounced.js (hunk 11 above)
Comment (basti1302) [reply packages/collector/src/announceCycle/unannounced.js:182]: I like this idea! I'll look into that.

File: packages/core/test/tracing/index_test.js (hunk 9 above)
Comment (basti1302) [reply packages/core/test/tracing/index_test.js:43]: discussed offline

File: packages/collector/src/announceCycle/unannounced.js (hunk 11 above)
Comment (basti1302) [reply packages/collector/src/announceCycle/unannounced.js]: I have implemented this suggestion now with the small difference that I attached the configuration that the agent has provided to agent/opts, which already was used to store the agent uuid, so it felt somewhat natural to put the config there, too.

I know you already approved but somehow this triggered a couple more changes (deleting an obsolete test) so I would actually appreciate a re-review of the latest changes if you don't mind :-)

File: packages/core/src/tracing/instrumentation/protocols/http2Server.js [hunk 12]
```diff
@@ -31,6 +31,26 @@ exports.init = function init(config) {
   extraHttpHeadersToCapture = config.tracing.http.extraHttpHeadersToCapture;
 };
 
+exports.updateConfig = function updateConfig(config) {
```
Comment (kirrg001) [inline packages/core/src/tracing/instrumentation/protocols/http2Server.js:34]: I'd like to talk about the `preinit` logic when you are back :)

File: packages/core/src/tracing/instrumentation/protocols/http2Server.js [hunk 13]
```diff
@@ -31,6 +31,26 @@ exports.init = function init(config) {
...
+  extraHttpHeadersToCapture = config.tracing.http.extraHttpHeadersToCapture;
+};
+
+exports.activate = function activate(extraConfig) {
+  if (
+    extraConfig &&
+    extraConfig.tracing &&
+    extraConfig.tracing.http &&
```
Comment (kirrg001) [inline packages/core/src/tracing/instrumentation/protocols/http2Server.js:42]: Suggestion: Sad we cannot use optional chaining. But maybe `utils.get(extraConfig, 'tracing.http')` (see https://lodash.com/docs#get)

**Review by kirrg001 — APPROVED** [review]
🥳

File: packages/core/src/tracing/instrumentation/protocols/http2Server.js (hunk 13 above)
Comment (basti1302) [reply packages/core/src/tracing/instrumentation/protocols/http2Server.js:42]: Yes, optional chaining would be great here but it is only available once we drop Node < 14.

TBH, I find the extra boiler plate okay-ish here. We can revisit the utils.get idea later, for now I would not go back and introduce a util for that here.
